File: src/main/java/org/nixos/idea/settings/NixExternalFormatterSettings.java
package org.nixos.idea.settings;

import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.components.PersistentStateComponent;
import com.intellij.openapi.components.RoamingType;
import com.intellij.openapi.components.State;
import com.intellij.openapi.components.Storage;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayDeque;
import java.util.Collection;
import java.util.Collections;
import java.util.Deque;

@State(name = "NixExternalFormatterSettings", storages = @Storage(value = NixStoragePaths.TOOLS, roamingType = RoamingType.LOCAL))
public final class NixExternalFormatterSettings implements PersistentStateComponent<NixExternalFormatterSettings.State> {

    // Documentation:
    // https://plugins.jetbrains.com/docs/intellij/persisting-state-of-components.html

    private static final int MAX_HISTORY_SIZE = 5;

    private @NotNull State myState = new State();

    public static @NotNull NixExternalFormatterSettings getInstance() {
        return ApplicationManager.getApplication().getService(NixExternalFormatterSettings.class);
    }

    public boolean isFormatEnabled() {
        return myState.enabled;
    }

    public void setFormatEnabled(boolean enabled) {
        myState.enabled = enabled;
    }

    public @NotNull String getFormatCommand() {
        return myState.command;
    }

    public void setFormatCommand(@NotNull String command) {
        myState.command = command;
        addFormatCommandToHistory(command);
    }

    public @NotNull Collection<String> getCommandHistory() {
        return Collections.unmodifiableCollection(myState.history);
    }

    private void addFormatCommandToHistory(@NotNull String command) {
        Deque<String> history = myState.history;
        history.remove(command);
        history.addFirst(command);
        while (history.size() > MAX_HISTORY_SIZE) {
            history.removeLast();
        }
    }

    @SuppressWarnings("ClassEscapesDefinedScope")
    @Override
    public void loadState(@NotNull State state) {
        myState = state;
    }

    @SuppressWarnings("ClassEscapesDefinedScope")
    @Override
    public @NotNull State getState() {
        return myState;
    }

    static final class State {
        public boolean enabled = false;
        public @NotNull String command = "";
        public Deque<String> history = new ArrayDeque<>();
    }
}

File: src/main/java/org/nixos/idea/settings/NixExternalFormatterSettings.kt
package org.nixos.idea.settings

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.components.BaseState
import com.intellij.openapi.components.RoamingType
import com.intellij.openapi.components.SimplePersistentStateComponent
import com.intellij.openapi.components.State
import com.intellij.openapi.components.Storage
import org.nixos.idea.settings.SimplePersistentStateComponentHelper.delegate
import java.util.ArrayDeque
import java.util.Collections
import java.util.Deque

@State(
    name = "NixExternalFormatterSettings",
    storages = [Storage(value = NixStoragePaths.TOOLS, roamingType = RoamingType.LOCAL)]
)
class NixExternalFormatterSettings : SimplePersistentStateComponent<NixExternalFormatterSettings.State>(State()) {

    class State : BaseState() {
        // Common options
        var enabled by property(false)
        var strategy by enum<Strategy>() // initialized by loadState() or noStateLoaded() for legacy purposes

        // Custom command configuration
        var command by string()
        var history: Deque<String> by property(ArrayDeque()) { it.isEmpty() }
        var inputMode by enum(InputMode.STDIO)
        var runInFileDirectory by property(false)
    }

    // Common options
    var isEnabled: Boolean by delegate(State::enabled)
    var strategy: Strategy
        get() = state.strategy!! // never null, always set by either loadState() or noStateLoaded()
        set(value) {
            state.strategy = value
        }

    // Custom command configuration
    var formatCommand: String by delegate(State::command, State::history)
    val formatCommandHistory: Collection<String>
        get() = Collections.unmodifiableCollection(state.history)
    var formatCommandInputMode: InputMode by delegate(State::inputMode)
    var isFormatCommandRunInFileDirectory: Boolean by delegate(State::runInFileDirectory)

    override fun noStateLoaded() {
        super.noStateLoaded()
        state.strategy = Strategy.NIX_FMT
        // New default for `runInFileDirectory` in new installations,
        // without changing the behavior of configurations written by older versions.
        state.runInFileDirectory = true
    }

    override fun loadState(state: NixExternalFormatterSettings.State) {
        super.loadState(state)
        if (state.strategy == null) {
            state.strategy = if (state.command.isNullOrBlank()) Strategy.NIX_FMT else Strategy.CUSTOM_COMMAND
        }
    }

    companion object {
        @JvmStatic
        fun getInstance(): NixExternalFormatterSettings {
            return ApplicationManager.getApplication().getService(NixExternalFormatterSettings::class.java)
        }
    }

    enum class Strategy {
        NIX_FMT,
        CUSTOM_COMMAND,
    }

    enum class InputMode {
        STDIO,
        CLI_ARGUMENT,
    }
}
